src/adapter/src/coord.rs

```diff
@@ -1577,6 +1589,86 @@ impl<S: Append + 'static> Coordinator<S> {
         }
     }
 
+    async fn message_advance_timelines(&mut self) {
+        // Convince the coordinator it needs to open a new timestamp
+        // and advance inputs.
+        // Fast forwarding puts the `TimestampOracle` in write mode,
+        // which means the next read may have to wait for table
+        // advancements. To prevent this we explicitly put
+        // the `TimestampOracle` in read mode. Writes will always
+        // advance a table no matter what mode the `TimestampOracle`
+        // is in. We step back the value of `now()` so that the
+        // next write can happen at `now()` and not a value above
+        // `now()`
+        let timelines: Vec<_> = self.global_timelines.keys().cloned().collect();
+        let nows: HashMap<_, _> = timelines
+            .into_iter()
+            .map(|timeline| {
+                let now = if timeline == Timeline::EpochMilliseconds {
+                    let now = self.now();
+                    now.step_back().unwrap_or(now)
+                } else {
+                    // For non realtime sources, we define now as the largest timestamp, not in
+                    // advance of any object's upper. This is the largest timestamp that is closed
+                    // to writes.
+                    let id_bundle = self.ids_in_timeline(&timeline);
+                    let mut now = Timestamp::minimum();
+                    let upper = self.least_valid_write(&id_bundle);
+                    let upper = if let Some(upper) = upper.as_option() {
+                        upper.step_back().unwrap_or_else(Timestamp::minimum)
+                    } else {
+                        Timestamp::MAX
+                    };
+                    now.join_assign(&upper);
+                    now
+                };
+                (timeline, now)
+            })
+            .collect();
+
+        for (timeline, timestamp_oracle) in &mut self.global_timelines {
+            let now = nows[timeline];
+            timestamp_oracle
+                .fast_forward(now, |ts| {
+                    self.catalog.persist_timestamp(timeline.to_string(), ts)
+                })
+                .await;
+            let _ = timestamp_oracle.read_ts();
+        }
+    }
+
+    fn ids_in_timeline(&self, timeline: &Timeline) -> CollectionIdBundle {
```

Comment:
Is it worth it saving this somewhere in the state of Coord or just re-computing it every time?